packages/x-charts/src/BarChart/BarPlot.tsx

```diff
@@ -199,15 +196,45 @@ const useAggregatedData = (): CompletedBarData[] => {
           width: verticalLayout ? barWidth : maxValueCoord - minValueCoord,
           color: colorGetter(dataIndex),
           highlightScope: series[seriesId].highlightScope,
+          value: series[seriesId].data[dataIndex],
+          maskId: `${chartId}_${series[seriesId].stack ?? ''}_${groupIndex}_${dataIndex}`,
         };
+
+        if (!masks[result.maskId]) {
+          masks[result.maskId] = {
+            id: result.maskId,
+            width: 0,
+            height: 0,
+            hasNegative: false,
+            hasPositive: false,
+            layout: result.layout,
+            xOrigin: xScale(0)!,
+            yOrigin: yScale(0)!,
+            x: 0,
+            y: 0,
+          };
+        }
+
+        const mask = masks[result.maskId];
+        mask.width = result.layout === 'vertical' ? result.width : mask.width + result.width;
+        mask.height = result.layout === 'vertical' ? mask.height + result.height : result.height;
+        mask.x = Math.min(mask.x === 0 ? Infinity : mask.x, result.x);
+        mask.y = Math.min(mask.y === 0 ? Infinity : mask.y, result.y);
+        mask.hasNegative = mask.hasNegative || (result.value ?? 0) < 0;
+        mask.hasPositive = mask.hasPositive || (result.value ?? 0) > 0;
+
+        return result;
       });
     });
   });
 
-  return data;
+  return {
+    completedData: data,
+    masksData: Object.values(masks),
+  };
 };
 
-const getOutStyle = ({ layout, yOrigin, x, width, y, xOrigin, height }: CompletedBarData) => ({
+const getOutStyle = ({ layout, yOrigin, x, width, y, xOrigin, height }: AnimationData) => ({
```

Comment:
Nitpick: The naming of this is a bit weird to me 🤔 I would go for `enterStyle` or `leaveStyle` 🙈 But if there is a specific reasoning here, feel free to ignore this

Reply:
It was like that, so I just left them there. Will change the name as I too found it a bit confusing at times 😅